src/managers/api.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for for handling API access."""
+import json
+import logging
+from typing import TYPE_CHECKING, Any
+
+import requests
+from requests.exceptions import RequestException
+
+from exceptions import OSDAPIError
+
+if TYPE_CHECKING:
+    pass
+
+from core.cluster import SUBSTRATES, ClusterState
+from core.workload import WorkloadBase
+
+logger = logging.getLogger(__name__)
+
+HEADERS = {
+    "Accept": "application/json",
+    "Content-Type": "application/json",
+    "osd-xsrf": "osd-true",
+}
+
+
+class APIManager:
+    """Manager for for handling configuration building + writing."""
+
+    def __init__(
+        self,
+        state: ClusterState,
+        workload: WorkloadBase,
+        substrate: SUBSTRATES,
+    ):
+        self.state = state
+        self.workload = workload
+        self.substrate = substrate
+
+    # =================================
+    #  Opensearch connection functions
+    # =================================
+
+    def request(
+        self,
+        endpoint: str,
+        method: str = "GET",
+        headers: dict = HEADERS,
+        payload: dict[str, Any] | None = None,
+    ) -> dict[str, Any]:
+        """Make an HTTP(S) request to the OSD Rest API.
+
+        Args:
+            method: matching the known http methods.
+            headers: request headers as a dict
+            endpoint: relative to the base uri.
+            payload: JSON / map body payload.
+        """
+
+        if None in [endpoint, method]:
+            raise ValueError("endpoint or method missing")
+
+        full_url = f"{self.state.unit_server.url}/api/{endpoint}"
+
+        request_kwargs = {
+            "verify": self.workload.paths.ca,
+            "method": method.upper(),
+            "url": full_url,
+            "headers": headers,
+        }
+
+        request_kwargs["data"] = json.dumps(payload)
+        request_kwargs["headers"] = headers
+
+        if not self.state.opensearch_server:
+            raise OSDAPIError("No Opensearch connection, can't query API (missing credentials).")
+
+        try:
+            with requests.Session() as s:
+                s.auth = (  # type: ignore [reportAttributeAccessIssue]
+                    self.state.opensearch_server.username,
+                    self.state.opensearch_server.password,
+                )
+                resp = s.request(**request_kwargs)
+                resp.raise_for_status()
+        except RequestException as e:
+            logger.error(f"Request {method} to {full_url} with payload: {payload} failed. \n{e}")
```

Comment:
I suggest raising a specific exception here, so you later in your code, you can watch for some few only exceptions in order to cover the entire code:
```suggestion
            logger.error(f"Request {method} to {full_url} with payload: {payload} failed. \n{e}")
            raise ODSHttpException(
                response_text=ex.response.text,
                response_code=ex.response.status_code,
            )  # composed of status code error and any message
```

Reply:
I thought of that option, and after some thinkgin/coding I purposefully chose that `self.request()` would return Python `requests` errors. (Generally I also rather prefer to have the approach you're describing.)

I quite like how Python `requests` groups its own exceptions under RequestException (much alongside the same consideration as what you are raising), and I prefer to distinguish later -- based on whther `self.requests()` returned a `ConnectionError` or a `HttpError` or an `SSLError` (all descendats of `RequestException`).

Typically if we want to further refine healthchecks or so -- we would like to distinguish between those, and not just hiding them (and reverse envineering them) from under a local exception's umbrella.

Alternatively of course we can also say that ANY of those result should in a `blocked` service... Yet we can't even devermine a charm status message in an elegant way if we do that.

Reply:
(Note that I'm purposefully using `repose.raise_for_status()` to have any HTTP errors bubbled up via `RequestException` for potential further analysis.)

Reply:
Ok, then please make it clear on the docstring of this method which exceptions should we expect and why

Reply:
Docstrings are now representing correctly the exceptions that may be thrown.

Closing thread.